### PR TITLE
Feature/admin page

### PR DESCRIPTION
**Pull request by RayMet**

### CODE REVIEW

File: client/src/components/AdminPage.jsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import React, { Component } from 'react';
+import { Link } from 'react-router-dom';
+
+
+class AdminPage extends Component {
```
Comment (Merteg) [inline client/src/components/AdminPage.jsx]: If you don’t have state or refs, prefer normal functions.[Airbnb Style Guide](https://github.com/airbnb/javascript/tree/master/react#class-vs-reactcreateclass-vs-stateless)

File: client/src/components/AdminPage.jsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  render() {
+    return (
+      <React.Fragment>
+        <Link to="/admin-page/users" className="btn waves-effect waves-light btn-flat">Users</Link>
+        <Link to="/" className="btn waves-effect waves-light btn-flat">Events</Link>
```
Comment (Merteg) [inline client/src/components/AdminPage.jsx]: This is link to home page, why did you call this `Events`?

File: client/src/components/AdminPageUsers.jsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    roles: [],
+    statuses: []
+  };
+
+  componentDidMount() {
+    request('/admin-page/users')
+      .then(data => {
+        this.setState({ users: data.users_dict });
+        this.setState({ roles: data.roles_dict})
+        this.setState({ statuses: data.statuses_dict})
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: You can use one setState instead of three.

File: client/src/components/AdminPageUsers.jsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      })
+  }
+
+  changeUserRole = (e) => {
+    let { id } = e.currentTarget;
+    id = id.slice(10);
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: It's can be hard to understand what `10` means, you can use here `id = id.replace(/user-role-/g, '')`

File: client/src/components/AdminPageUsers.jsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    const data = {
+      "role_id": e.currentTarget.value
+    }
+    request(`/admin-page/users/${id}`, "PUT", JSON.stringify(data))
+      .then(data => {
+        if ('role_id' in data) {
+        let users_arr = this.state.users;
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: Wrong indentation.

File: client/src/components/AdminPageUsers.jsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        this.setState({
+          users: users_arr
+        })
+        }
+      })
+  }
+  changeUserStatus = (e) => {
+    let { id } = e.currentTarget;
+    id = id.slice(5);
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: Also better to use `.replace`

File: client/src/components/AdminPageUsers.jsx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    )
+  }
+
+  renderStatuslist() {
+    return (
+      <React.Fragment>
+        {this.state.statuses.map((element) => {
+          if (element.status !== 'Non_active') {
+            return (<option value={element.id}>{element.status}</option>);
+          } else return 0;
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: You don't need this `return 0`

File: client/src/components/HomePage/LogIn.jsx [hunk 8]
```diff
@@ -36,6 +36,7 @@ class LogIn extends Component {
           sessionStorage.setItem("User-nickname", data.user.nickname);
           sessionStorage.setItem("User-avatar", data.user.avatar);
           sessionStorage.setItem("User-id", data.user.user_id);
+          sessionStorage.setItem("User-role", data.user.user_role);
```
Comment (Merteg) [inline client/src/components/HomePage/LogIn.jsx:39]: Where do you use this

File: client/src/components/Routes.jsx [hunk 9]
```diff
@@ -13,13 +15,19 @@ import EventPage from './EventPage';
 import ProfileEdit from './ProfileEdit';
 
 
-function Routes() {
+
+class Routes extends Component {
```
Comment (Merteg) [inline client/src/components/Routes.jsx]: Why you change it to class?

File: server/server/views/admin.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    def collection_get(self):
+        users_list = User.get_all(self.request)
+        roles_list = Role.get_all(self.request)
+        statuses_list = UserStatus.get_all(self.request)
+        users_dict = []
+        roles_dict = []
+        statuses_dict = []
+        for obj in users_list:
+            temp_dict = model_to_dict(obj)
+            temp_dict['status_str'] = obj.user_statuses.status
```
Comment (Merteg) [inline server/server/views/admin.py]: You have all statuses in statuses_dict, so you don't need to add a status_str to user dict.

File: server/server/views/admin.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            users_dict.append(temp_dict)
+        roles_dict = [model_to_dict(obj) for obj in roles_list]
+        statuses_dict = [model_to_dict(obj) for obj in statuses_list]
+        response = {}
```
Comment (Merteg) [inline server/server/views/admin.py]: You can add arrays to response when you define it. Also, you can immediately return it, without a response variable.

File: server/server/views/admin.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    @view(permission="admin")
+    def put(self):
+        request = self.request
+        json = request.json_body
+        user = User.get_one(request, id=request.matchdict['id'])
+        if user is not None:
+            if 'role_id' in json:
+                user.role_id = json['role_id']
+                return {'role_id': json['role_id']}
+            if 'status_id' in json and 'status':
```
Comment (Merteg) [inline server/server/views/admin.py]: Emmm, `... and 'status'`. What is it?

File: server/server/views/auth.py [hunk 13]
```diff
@@ -36,7 +36,8 @@ def log_in_post(request):
             response['user'] = {
                 'nickname': user.nickname,
                 'avatar': user.avatar,
-                'user_id': user.id
+                'user_id': user.id,
+                'user_role': user.role_id
```
Comment (Merteg) [inline server/server/views/auth.py]: You don't use it on the client, so you don't need it here.

**Review by Merteg — CHANGES_REQUESTED** [review]

File: server/server/views/admin.py (hunk 10 above)
Comment (RayMet) [reply server/server/views/admin.py]: I made this because, don't want maps inside maps in AdminPageUsers

File: server/server/views/admin.py (hunk 11 above)
Comment (RayMet) [reply server/server/views/admin.py]: How can I do that?

File: client/src/components/AdminPageUsers.jsx (hunk 4 above)
Comment (RayMet) [reply client/src/components/AdminPageUsers.jsx]: What is "g"?

**Comment by RayMet** [comment]
Review please

**Comment by RayMet** [comment]
Update

File: client/src/components/AdminPage.jsx [hunk 14]
```diff
@@ -1,17 +1,14 @@
...
-    return (
-      <React.Fragment>
-        <Link to="/admin-page/users" className="btn waves-effect waves-light btn-flat">Users</Link>
-        <Link to="/" className="btn waves-effect waves-light btn-flat">Events</Link>
-      </React.Fragment>
+const AdminPage = () => {
+  return (
+    <React.Fragment>
+      <Link to="/admin-page/users" className="btn waves-effect waves-light btn-flat">Users</Link>
+      <Link to="/admin-page/events" className="btn waves-effect waves-light btn-flat">Events</Link>
```
Comment (Merteg) [inline client/src/components/AdminPage.jsx]: This link doesn't match anything now, why did you add it?

File: client/src/components/AdminPageUsers.jsx [hunk 15]
```diff
@@ -18,12 +19,20 @@ class AdminPageUsers extends Component {
...
+        );
+      } else return 0;
+
+    })
   }
 
   changeUserRole = (e) => {
     let { id } = e.currentTarget;
-    id = id.slice(10);
+    id = id.replace(/user-role-/g, '')
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: You can use `id = id.replace('user-role-', '')`, seems really better.

**Comment by Merteg** [comment]
Also, your admin doesn't work in the right way, user statuses don't work.
You can test your code before pushing it, it's a good practice.

**Comment by Merteg** [comment]
oh, a few more...
image
Remove this `Materialize Select` please.
Also, in my opinion, better to show some message when status/role was changed successfully.

**Comment by RayMet** [comment]
New update

File: server/server/views/admin.py [hunk 16]
```diff
@@ -23,6 +23,8 @@ def collection_get(self):
         users_list = User.get_all(self.request)
         roles_list = Role.get_all(self.request)
         statuses_list = UserStatus.get_all(self.request)
+        Non_active_status_id_dict = UserStatus\
```
Comment (Merteg) [inline server/server/views/admin.py]: First, it's not dict. Second, why the first letter is uppercase?

File: client/src/components/AdminPageUsers.jsx [hunk 17]
```diff
@@ -8,7 +8,8 @@ class AdminPageUsers extends Component {
     users: [],
     roles: [],
     statuses: [],
-    not_active_id: ""
+    non_active_id: 0,
+    change_color: ""
```
Comment (Merteg) [inline client/src/components/AdminPageUsers.jsx]: You never use this

**Review by Merteg — COMMENTED** [review]
First, this design is bad, but it's at your discretion.
image
Second, add a redirect on admin page after login or link to the nav, why the user should type `admin-page` every time?

**Comment by Merteg** [comment]
image
Also, you have warnings at your admin page, please resolve.

**Comment by RayMet** [comment]
Update

**Review by Merteg — APPROVED** [review]

**Review by ykobuk — APPROVED** [review]
